subprojects/logging/src/main/java/org/gradle/internal/logging/sink/OutputEventRenderer.java

```diff
@@ -290,30 +292,30 @@ public void configureMaxWorkerCount(int maxWorkerCount) {
 
     @Override
     public void onOutput(OutputEvent event) {
-        synchronized (lock) {
-            if (event.getLogLevel() != null && event.getLogLevel().compareTo(logLevel) < 0) {
+        if (event.getLogLevel() != null && event.getLogLevel().compareTo(logLevel.get()) < 0) {
+            return;
+        }
+        if (event instanceof LogLevelChangeEvent) {
+            LogLevelChangeEvent changeEvent = (LogLevelChangeEvent) event;
+            LogLevel newLogLevel = changeEvent.getNewLogLevel();
+            if (newLogLevel == this.logLevel.get()) {
                 return;
             }
-            if (event instanceof LogLevelChangeEvent) {
-                LogLevelChangeEvent changeEvent = (LogLevelChangeEvent) event;
-                LogLevel newLogLevel = changeEvent.getNewLogLevel();
-                if (newLogLevel == this.logLevel) {
-                    return;
-                }
-                this.logLevel = newLogLevel;
-            } else if (event instanceof MaxWorkerCountChangeEvent) {
-                MaxWorkerCountChangeEvent changeEvent = (MaxWorkerCountChangeEvent) event;
-                int newMaxWorkerCount = changeEvent.getNewMaxWorkerCount();
-                if (newMaxWorkerCount == this.maxWorkerCount) {
-                    return;
-                }
-                this.maxWorkerCount = newMaxWorkerCount;
+            this.logLevel.set(newLogLevel);
+        } else if (event instanceof MaxWorkerCountChangeEvent) {
+            MaxWorkerCountChangeEvent changeEvent = (MaxWorkerCountChangeEvent) event;
+            int newMaxWorkerCount = changeEvent.getNewMaxWorkerCount();
+            if (newMaxWorkerCount == this.maxWorkerCount.get()) {
+                return;
             }
+            this.maxWorkerCount.set(newMaxWorkerCount);
+        }
+        synchronized (lock) {
```

Comment:
⭕️  Could this be in an else branch or do the formatters need to know about level/max worker change events?